developer/src/kmcmplib/src/CompilerInterfaces.cpp

```diff
@@ -0,0 +1,368 @@
+    SetError(CERR_InfileNotExist);
+  }
+
+  // Transfer the file to a memory stream for processing UTF-8 or ANSI to UTF-16?
+  // What about really large files?  Transfer to a temp file...
+  if (!fread(str, 1, 3, fp_in)) {
+    fclose(fp_in);
+    SetError(CERR_CannotReadInfile);
+  }
+
+  fseek(fp_in, 0, SEEK_SET);
+  if (str[0] == UTF8Sig[0] && str[1] == UTF8Sig[1] && str[2] == UTF8Sig[2])
+    fp_in = UTF16TempFromUTF8(fp_in, TRUE);
+  else if (str[0] == UTF16Sig[0] && str[1] == UTF16Sig[1])
+    fseek(fp_in, 2, SEEK_SET);
+  else
+    fp_in = UTF16TempFromUTF8(fp_in, FALSE);
+  if (fp_in == NULL) {
+    SetError(CERR_CannotCreateTempfile);
+  }
+
+  fp_out = Open_File(pszOutfile, "wb");
+
+  if (fp_out == NULL) {
+    SetError(CERR_CannotCreateOutfile);
+  }
+
+  FILE_KEYBOARD fk;
+  kmcmp::CodeConstants = new kmcmp::NamedCodeConstants;
+  bool result = CompileKeyboardHandle(fp_in, &fk);
+  if(result) {
+    KMX_DWORD msg;
+    if ((msg = WriteCompiledKeyboard(&fk, fp_out)) != CERR_None) {
+      result = FALSE;
+      AddCompileError(msg);
+    }
+  } else {
+    AddCompileError(CERR_InvalidValue);
+  }
+
+  fclose(fp_in);
+  fclose(fp_out);
+
+  delete kmcmp::CodeConstants;
+
+  if (kmcmp::nErrors > 0)
+  {
+    remove(pszOutfile);
+    return FALSE;
+  }
+
+  return result;
+}
+
+
+
+EXTERN bool kmcmp_CompileKeyboardFileToBuffer(char* pszInfile, void* pfkBuffer, bool ACompilerWarningsAsErrors, bool AWarnDeprecatedCode,
+  kmcmp_CompilerMessageProc pMsgproc, void* AmsgprocContext, int Target)   // I4865   // I4866
+{
+  //printf("°°-> changed to CompileKeyboardFileToBuffer() of kmcmplib \n");
+  FILE* fp_in = NULL;
+  KMX_CHAR str[260];
+
+  kmcmp::FSaveDebug = TRUE;   // I3681
+  kmcmp::FCompilerWarningsAsErrors = ACompilerWarningsAsErrors;   // I4865
+  AWarnDeprecatedCode_GLOBAL_LIB = AWarnDeprecatedCode;
+  kmcmp::CompileTarget = Target;
+
+  if (!pMsgproc || !pszInfile || !pfkBuffer) {
+    SetError(CERR_BadCallParams);
+  }
+
+  PKMX_STR p;
+
+  if ((p = strrchr_slash(pszInfile)) != nullptr) {
+    strncpy(kmcmp::CompileDir, pszInfile, (int)(p - pszInfile + 1));  // I3481
+    kmcmp::CompileDir[(int)(p - pszInfile + 1)] = 0;
+  }
+  else {
+    kmcmp::CompileDir[0] = 0;
+  }
+
+  msgproc = pMsgproc;
+  msgprocContext = AmsgprocContext;
+  kmcmp::currentLine = 0;
+  kmcmp::nErrors = 0;
+
+  fp_in = Open_File(pszInfile,"rb");
+
+  if (fp_in == NULL) {
+    SetError(CERR_InfileNotExist);
+  }
+
+  // Transfer the file to a memory stream for processing UTF-8 or ANSI to UTF-16?
+  // What about really large files?  Transfer to a temp file...
+
+  if( !fread(str,1,3,fp_in))
+  {
+    fclose(fp_in);
+    SetError(CERR_CannotReadInfile);
+  }
+
+  fseek( fp_in,0,SEEK_SET);
+  if (str[0] == UTF8Sig[0] && str[1] == UTF8Sig[1] && str[2] == UTF8Sig[2])
+    fp_in = UTF16TempFromUTF8(fp_in, TRUE);
+  else if (str[0] == UTF16Sig[0] && str[1] == UTF16Sig[1])
+    fseek( fp_in,2,SEEK_SET);
+  else
+    fp_in = UTF16TempFromUTF8(fp_in, FALSE);
```

Comment:
Is there an opportunity to refactor this with ll. 117-122?

Reply:
Yes, separate PR though.